roles/translator/src/job_negotiator/mod.rs

```diff
@@ -0,0 +1,129 @@
+        let initiator = Initiator::from_raw_k(authority_public_key).unwrap();
+        let (mut receiver, mut sender): (Receiver<EitherFrame>, Sender<EitherFrame>) =
+            Connection::new(stream, HandshakeRole::Initiator(initiator), 10).await;
+
+        let proxy_address = SocketAddr::new(
+            IpAddr::from_str(&config.downstream_address).unwrap(),
+            config.downstream_port,
+        );
+
+        info!(
+            "JN proxy: setupconnection Proxy address: {:?}",
+            proxy_address
+        );
+
+        SetupConnectionHandler::setup(&mut receiver, &mut sender, proxy_address)
+            .await
+            .unwrap();
+
+        info!("JN CONNECTED");
+
+        let self_ = Arc::new(Mutex::new(JobNegotiator {
+            receiver,
+            _sender: sender,
+            sender_coinbase_output_max_additional_size,
+            last_coinbase_out: None,
+            sender_coinbase_out,
+            coinbase_reward_sat: config.coinbase_reward_sat,
+        }));
+
+        Self::on_upstream_message(self_.clone());
+        self_
+    }
+
+    pub fn on_upstream_message(self_mutex: Arc<Mutex<Self>>) {
+        async_std::task::spawn(async move {
+            let sender_max_size = self_mutex
+                .safe_lock(|s| s.sender_coinbase_output_max_additional_size.clone())
+                .unwrap();
+            let sender_out_script = self_mutex
+                .safe_lock(|s| s.sender_coinbase_out.clone())
+                .unwrap();
+            let receiver = self_mutex.safe_lock(|d| d.receiver.clone()).unwrap();
+            loop {
+                let mut incoming: StdFrame = receiver.recv().await.unwrap().try_into().unwrap();
+                let message_type = incoming.get_header().unwrap().msg_type();
+                let payload = incoming.payload();
+
+                let next_message_to_send =
+                    ParseServerJobNegotiationMessages::handle_message_job_negotiation(
+                        self_mutex.clone(),
+                        message_type,
+                        payload,
+                    );
+                match next_message_to_send {
+                    Ok(SendTo::None(Some(JobNegotiation::SetCoinbase(m)))) => {
+                        let coinbase_output_max_additional_size = CoinbaseOutputDataSize {
+                            coinbase_output_max_additional_size: m
+                                .coinbase_output_max_additional_size,
+                        };
+                        let out_script = self_mutex
+                            .safe_lock(|s| s.last_coinbase_out.clone())
+                            .unwrap()
+                            // Safe unwrap when we receive a coinbase_output we immediatly set
+                            // last_coinbase_out to that coinbase output in the message handlers
+                            // message_handler::ParseServerJobNegotiationMessages::handle_set_coinbase
+                            .unwrap();
```

Comment:
:( - we should get in the habit of using `handle_result` and the tx_status channel to report errors up to the main thread.

Reply:
For example - https://github.com/stratum-mining/stratum/blob/main/roles/translator/src/upstream_sv2/upstream.rs#L252

Reply:
Here we do not want to use `handle_result`. We must always set `self.last_coinbase_out` to `Some(coinbase_outs)` before execute the above code. If this is not done it means that the code is bugged and must be fixed is not an error that we should handle somewhere else.

Reply:
That is the point of handle_result - as the code is now if the unwrap() fails the task will fail and the translator will be in a degraded state and will no longer be handling or sending JN messages. By adding the `handle_result` mechanism if there is a failure the tproxy will die and it'll be apparent that there was a bug.

Reply:
IMO no. 

`handle_result` should not be used as a debug tool. We must use `handle_result` in all that cases where we need to pass the error cause we can not handle the error there and we can took some actions elsewhere. Here we have a condition that if true means that the code of the tproxy is bugged. It can not arise cause a peer is sending malformed messages or is not responding ecc ecc. If the above happen means that the logic of the tproxy is wrong this is a bug  and must be solved not an error that can arise and should be handled.

Note that all this reasoning is not about this specific case but general one, cause we are changing the `SetCoinbase` message anyway. 

In this case when we receive the `SetCoinbase` message the tproxy must set `self.last_coinbase_out` to the received value so it is impossible to have None value here. If we have a None value here it means that the code is bugged we are in a state that should always be impossible to reach so it make no sense to have an handler for that. We just panic and as developer we see that the code do panic here and we fix the bug.

Does it make sense?